### PR TITLE
refac: add Digest traits

### PR DESCRIPTION
**Pull request by Insun35**
This PR introduces `Digest` traits for `Sha256Targets` to digest data successively in a chained manner. The main Sha256 gadget logic is refactored into the following inline functions:
- `new()`: construct hasher with empty data
- `update()`: update input message
- `chain_update()`: update in a chained manner
- `finalize()`: return the digested data

### CODE REVIEW

**Review by silathdiir — APPROVED** [review]
LGTM

File: src/circuit.rs [hunk 1]
```diff
@@ -4,69 +4,172 @@ use plonky2::iop::target::BoolTarget;
...
+    /// Retrieve result and consume hasher instance.
+    fn finalize(self, builder: &mut CircuitBuilder<F, D>) -> Sha256Targets;
+}
+
+impl<F, const D: usize> Digest<F, D> for Sha256Targets
+where
+    F: RichField + Extendable<D>,
+{
+    #[inline]
+    fn new(builder: &mut CircuitBuilder<F, D>) -> Self {
```
Comment (silathdiir) [inline src/circuit.rs]: Seems the parameter `builder` is useless?

**Review by silathdiir — APPROVED** [review]
Left a small comment, others LGTM!

File: src/circuit.rs (hunk 1 above)
Comment (Insun35) [reply src/circuit.rs]: Yep, you are right. And also I think we don't need to separate functions in each `Update` and `Digest` traits, so I just removed `Update` traits and unified to `Digest` traits. Thanks!

**Review by nikkolasg — APPROVED** [review]

**Pull request merged**